msal/broker.py

```diff
@@ -0,0 +1,198 @@
+        self.auth_result = auth_result
+
+
+def _convert_error(error, client_id):
+    context = error.get_context()  # Available since pymsalruntime 0.0.4
+    if "AADSTS50011" in context:  # In WAM, this could happen on both interactive and silent flows
+        raise RedirectUriError(  # This would be seen by either the app developer or end user
+            "MsalRuntime won't work unless this one more redirect_uri is registered to current app: "
+            "ms-appx-web://Microsoft.AAD.BrokerPlugin/{}".format(client_id))
+    return {
+        "error": "broker_error",  # Note: Broker implies your device needs to be compliant.
+            # You may use "dsregcmd /status" to check your device state
+            # https://docs.microsoft.com/en-us/azure/active-directory/devices/troubleshoot-device-dsregcmd
+        "error_description": "{}. Status: {}, Error code: {}, Tag: {}".format(
+            context,
+            error.get_status(), error.get_error_code(), error.get_tag()),
+        }
+
+
+def _read_account_by_id(account_id):
+    """Return the callback result which contains the account or error"""
+    callback_data = _CallbackData()
+    pymsalruntime.read_account_by_id(
+        account_id,
+        "correlation_id",
+        lambda result, callback_data=callback_data: callback_data.complete(result)
+        )
+    callback_data.signal.wait()
+    return callback_data.auth_result
+
+
+def _convert_result(result, client_id):  # Mimic an on-the-wire response from AAD
+    error = result.get_error()
+    if error:
+        return _convert_error(error, client_id)
+    id_token_claims = json.loads(result.get_id_token()) if result.get_id_token() else {}
+    account = result.get_account()
+    assert account, "Account is expected to be always available"
+    ## Note: As of pymsalruntime 0.1.0, only wam_account_ids property is available
+    #account.get_account_property("wam_account_ids")
+    return_value = {k: v for k, v in {
+        "access_token": result.get_access_token(),
+        "expires_in": result.get_access_token_expiry_time(),
+        "id_token_claims": id_token_claims,
+        "client_info": account.get_client_info(),
+        "_account_id": account.get_account_id(),
+        }.items() if v}
+    granted_scopes = result.get_granted_scopes()  # New in pymsalruntime 0.3.x
+    if granted_scopes:
+        return_value["scope"] = " ".join(granted_scopes)  # Mimic the on-the-wire data format
+    return return_value
+
+
+def _get_new_correlation_id():
+    return str(uuid.uuid4())
+
+
+def _signin_silently(authority, client_id, scopes, correlation_id=None, **kwargs):
+    params = pymsalruntime.MSALRuntimeAuthParameters(client_id, authority)
+    params.set_requested_scopes(scopes)
+    callback_data = _CallbackData()
+    for k, v in kwargs.items():  # This can be used to support domain_hint, max_age, etc.
+        if v is not None:
+            params.set_additional_parameter(k, str(v))
+    pymsalruntime.signin_silently(
```

Comment:
>pymsalruntime.signin_silently [](http://example.com/codeflow?start=4&length=29)

All operation functions will return a handle so that you can use that handle to call cancel.

Reply:
Thanks. Good to know. MSAL Python can consider using it when the need arises. So far, there seems to be unnecessary to add a cancel behavior into MSAL (Python)'s API surface: broker's silent operations seem reasonably responsive, and its interactive operations have a Close button in its pop-up window, anyway.

Reply:
Any operations that will hit the wire has potential latency, even for the silent request. I will let you decide when to add this feature but the functionality of returning a cancel for cancel operation is there.